src/ray/core_worker/core_worker.cc

```diff
@@ -1193,6 +1241,24 @@ Status CoreWorker::ExecuteTask(const TaskSpecification &task_spec,
   return status;
 }
 
+Status CoreWorker::ExecuteTaskLocalMode(const TaskSpecification &task_spec,
+                                        const ActorID &actor_id) {
+  auto resource_ids = std::make_shared<ResourceMappingType>();
+  auto return_objects = std::vector<std::shared_ptr<RayObject>>();
+  auto borrowed_refs = ReferenceCounter::ReferenceTableProto();
+  for (size_t i = 0; i < task_spec.NumReturns(); i++) {
+    reference_counter_->AddOwnedObject(task_spec.ReturnId(i, TaskTransportType::DIRECT),
+                                       /*inner_ids=*/{}, GetCallerId(), rpc_address_,
+                                       CurrentCallSite(), -1);
+  }
+  auto old_id = GetActorId();
+  SetActorId(actor_id);
+  auto status = ExecuteTask(task_spec, resource_ids, &return_objects, &borrowed_refs);
```

Comment:
I don't think we can execute the task directly in the current thread. This will blocks user code (breaks the semantics that `func.remote()` is async). And it will also lead to dead locks in some situations.

Reply:
Local Mode should operate like serial python, as per its definition [here](https://github.com/ray-project/ray/blob/86e19959a5cbd91b32306a2e26481d9b2abe1f95/python/ray/worker.py#L224).